File: Full Stack Develop/Programming Language/Python/demos/d13_iterators.py
"""13 Iterators & generators — lazily produce sequences, computing only as much as you use.

Key points:
  1. A generator function uses yield to emit values one at a time; calling it returns a lazy iterator.
  2. A generator expression (x for x in ...) is the lazy version of a comprehension.
  3. Lazy: values are computed when pulled, so it can represent infinite sequences, truncated with itertools.islice.
  4. yield from delegates to another iterable, for concatenation/recursion.
  5. A custom iterator implements __iter__/__next__; for/spread/destructuring are all built on the iteration protocol.
"""

import itertools
from collections.abc import Iterable, Iterator

from .log import note, show, title


def gen_range(start: int, end: int, step: int = 1) -> Iterator[int]:
    """Generator function: written like an ordinary function, emitting values with yield."""
    i = start
    while i < end:
        yield i
        i += step


def fib() -> Iterator[int]:
    """Infinite sequence: lazy, so while True won't run away."""
    a, b = 0, 1
    while True:
        yield a
        a, b = b, a + b


def concat[T](*iterables: Iterable[T]) -> Iterator[T]:
    """yield from delegation: splice multiple iterables into one."""
    for it in iterables:
        yield from it


class Countdown:
    """Custom iterable: implementing __iter__ makes it consumable by for."""

    def __init__(self, start: int) -> None:
        self.start = start

    def __iter__(self) -> Iterator[int]:
        n = self.start
        while n > 0:
            yield n
            n -= 1


def run() -> None:
    title("13 Iterators & generators")

    note("generator function: consume with list() or for")
    show("list(gen_range(0, 5))", list(gen_range(0, 5)))
    show("list(gen_range(0, 10, 2))", list(gen_range(0, 10, 2)))

    note("infinite sequence + islice: lazy evaluation, take only the first 10")
    show("islice(fib(), 10)", list(itertools.islice(fib(), 10)))

    note("generator expression: the lazy version of a comprehension, () not []")
    squares = (n * n for n in fib())  # not computed immediately
    show("first 5 fib squares", list(itertools.islice(squares, 5)))

    note("yield from delegation: splice multiple iterables")
    show("concat(range, list)", list(concat(gen_range(0, 3), [100, 200])))

    note("custom iterable: implement __iter__")
    show("list(Countdown(5))", list(Countdown(5)))

    note("manual driving: next() pulls one at a time, raises StopIteration when exhausted")
    it = gen_range(0, 2)
    show("next(it)", next(it))
    show("next(it)", next(it))
    show("next(it, 'end')", next(it, "end"))  # a default avoids the exception

File: Full Stack Develop/Programming Language/Rust/src/demos/d13_iterators.rs
use crate::log::{note, show, title};

/// Iterators — Rust's lazy sequence abstraction; adapter chains are the daily workhorse (also the equivalent of "generators").
/// Key points:
///   1. The Iterator trait needs only next() implemented; adapters like map/filter/take are all lazy.
///   2. Lazy: the computation in a chain happens at collect/for, so it can represent infinite sequences.
///   3. A custom iterator = implementing Iterator for your own struct.
///   4. Adapters are zero-cost: after compilation they're as fast as a hand-written loop.

// custom infinite iterator: Fibonacci
struct Fib {
    a: u64,
    b: u64,
}
impl Iterator for Fib {
    type Item = u64;
    fn next(&mut self) -> Option<u64> {
        let cur = self.a;
        self.a = self.b;
        self.b = cur + self.b;
        Some(cur) // never returns None → infinite
    }
}

// custom finite iterator: countdown
struct Countdown {
    n: u32,
}
impl Iterator for Countdown {
    type Item = u32;
    fn next(&mut self) -> Option<u32> {
        if self.n == 0 {
            None // returning None ends it
        } else {
            self.n -= 1;
            Some(self.n + 1)
        }
    }
}

pub fn run() {
    title("13 Iterators");

    note("adapter chain: lazy, evaluated only at collect");
    let doubled: Vec<i32> = (1..=5).map(|x| x * 2).collect();
    show("(1..=5).map(*2)", &doubled);

    note("range + take: take only what you need from something infinite");
    let squares: Vec<i32> = (1..).map(|x| x * x).take(5).collect();
    show("(1..).map(sq).take(5)", &squares);

    note("custom infinite iterator Fib + take");
    let fibs: Vec<u64> = Fib { a: 0, b: 1 }.take(10).collect();
    show("Fib.take(10)", &fibs);

    note("custom finite iterator: next returning None ends it");
    let cd: Vec<u32> = Countdown { n: 5 }.collect();
    show("Countdown(5)", &cd);

    note("common adapters: filter / zip / fold / chain");
    let evens: Vec<i32> = (1..=10).filter(|x| x % 2 == 0).collect();
    show("filter evens", &evens);
    let zipped: Vec<(char, i32)> = ['a', 'b', 'c'].into_iter().zip(1..).collect();
    show("zip", &zipped);
    show("fold sum", (1..=5).fold(0, |acc, x| acc + x));
    let chained: Vec<i32> = (0..2).chain(10..12).collect();
    show("chain", &chained);
}
